// @codious/sav-reader-writer/utilities/buffer.js
// PRIVATE PROPERTIES
const types = Object.freeze({ Buffer: 'buffer', Float: 'float', Int8: 'int8', Int32: 'int32', String: 'string' });


// PUBLIC PROPERTIES
module.exports.types = types;


// PUBLIC METHODS
module.exports.write = function(data) {
  data = Array.isArray(data) ? data : [data]; // eslint-disable-line no-param-reassign
  return Buffer.concat(data.map((item) => { // eslint-disable-line array-callback-return, consistent-return
    switch (item.type) { // eslint-disable-line default-case
      case types.Buffer: return item.value;
      case types.Float: return writeFloat(item.value);
      case types.Int8: return writeInt8(item.value);
      case types.Int32: return writeInt32(item.value);
      case types.String: return writeString(item.value, item.length);
    }
  }));
};


// PRIVATE METHODS
function writeFloat(value) {
  const buffer = Buffer.alloc(8);
  buffer.writeDoubleLE(value);
  return buffer;
}

function writeInt8(value) {
  const buffer = Buffer.alloc(1);
  buffer.writeUInt8(value);
  return buffer;
}

function writeInt32(value) {
  const buffer = Buffer.alloc(4);
  buffer.writeInt32LE(value);
  return buffer;
}

function writeString(value, length) {
  if (length && value.length > length) value = value.slice(0, length); // eslint-disable-line no-param-reassign
  if (length && value.length < length) value = value.padEnd(length); // eslint-disable-line no-param-reassign
  return Buffer.from(value, 'utf8');
}
